qmcpy/integrand/barrier_option.py

```diff
@@ -38,42 +84,87 @@ def __init__(self, n = 4096, volatility=0.2, start_price=30, strike_price=35, ba
+            self.leveltype = 'fixed-multi'
+            self.parent = True
+            self.parameters += ['multilevel_dims']
+        else: # single level problem
+            self.dim_frac = _dim_frac
+            self.leveltype = 'single'
+            self.parent = False
+            self.parameters += ['dim_frac']
         super(BarrierOption,self).__init__(dimension_indv=1,dimension_comb=1,parallel=False)   
 
-    def get_discounted_payoff(self):
-        stock_values = self.start_price*exp((self.interest_rate-0.5*self.volatility**2)*self.t+self.volatility*self.true_measure.gen_samples(self.n))
-        print("Stock Values\n", stock_values)
-        expected_stock = stock_values[:,self.observations - 1]
-        print("Expected Stocks\n", expected_stock)
+    def _get_discounted_payoffs(self,stock_path,dimension):
+        """
+        Calculate the discounted payoff from the stock path. 
+        
+        Args:
+            stock_path (ndarray): n samples by d dimension option prices at monitoring times
+            dimension (int): number of dimensions
+        
+        Return:
+            ndarray: n vector of discounted payoffs
+        """
+        expected_stock = stock_path[:,dimension - 1]
         if(self.call):
             disc_payoff = expected_stock - self.strike_price
         else:
             disc_payoff = self.strike_price - expected_stock
-        print("Payoff without activation\n",disc_payoff)
         if(self.up and self.i):
-            bar_flag = stock_values >= self.barrier_price
-            print("Pre-barrier values\n",bar_flag)
+            bar_flag = stock_path >= self.barrier_price
             bar_flag = bar_flag.sum(axis = 1) > 0
         elif(self.up and (self.i is False)):
-            bar_flag = stock_values < self.barrier_price
-            print("Pre-barrier values\n",bar_flag)
-            bar_flag = bar_flag.sum(axis = 1) == self.observations
+            bar_flag = stock_path < self.barrier_price
+            bar_flag = bar_flag.sum(axis = 1) == dimension
         elif((self.up is False) and self.i):
-            bar_flag = stock_values <= self.barrier_price
-            print("Pre-barrier values\n",bar_flag)
+            bar_flag = stock_path <= self.barrier_price
             bar_flag = bar_flag.sum(axis = 1) > 0
         else:
-            bar_flag = stock_values > self.barrier_price
-            print("Pre-barrier values\n",bar_flag)
-            bar_flag = bar_flag.sum(axis = 1) == self.observations
-        print("Activating/Deactivation array\n",bar_flag)
+            bar_flag = stock_path > self.barrier_price
+            bar_flag = bar_flag.sum(axis = 1) == dimension
         disc_payoff = disc_payoff*bar_flag
-        print("Discounted payoff after activation\n", disc_payoff)
         disc_payoff = maximum(zeros(disc_payoff.size), disc_payoff)
-        print("Final Discounted Payoff vector\n", disc_payoff)
-        return disc_payoff.mean()
+        return disc_payoff
```

Comment:
I think you need to discount this e.g. `disc_payoff = disc_payoff * exp(-self.interest_rate * self.t_final)`.  @fredhickernell, can you take a look?